### PR TITLE
Fix resource quota shutdown

### PR DESCRIPTION
**Pull request by wojtek-t**
Ref https://github.com/kubernetes/kubernetes/issues/108483

```release-note
NONE
```

/kind cleanup
/priority important-longterm
/sig auth

### CODE REVIEW

File: staging/src/k8s.io/apiserver/pkg/server/options/admission.go [hunk 1]
```diff
@@ -143,9 +143,8 @@ func (a *AdmissionOptions) ApplyTo(
 	if err != nil {
 		return err
 	}
-	genericInitializer := initializer.New(clientset, informers, c.Authorization.Authorizer, features)
-	initializersChain := admission.PluginInitializers{}
-	pluginInitializers = append(pluginInitializers, genericInitializer)
+	genericInitializer := initializer.New(clientset, informers, c.Authorization.Authorizer, features, c.DrainedNotify())
+	initializersChain := admission.PluginInitializers{genericInitializer}
```
Comment (wojtek-t) [inline staging/src/k8s.io/apiserver/pkg/server/options/admission.go:147]: This is on purpose changing the order of plugin initializers.

Previously the one setting QuotaEvaluator:
https://github.com/kubernetes/kubernetes/blob/master/pkg/kubeapiserver/admission/initializer.go#L64
was called before the genericInitializer that was setting stopCh added in this PR.

File: staging/src/k8s.io/apiserver/pkg/admission/initializer/initializer.go [hunk 2]
```diff
@@ -39,12 +40,14 @@ func New(
 	extInformers informers.SharedInformerFactory,
 	authz authorizer.Authorizer,
 	featureGates featuregate.FeatureGate,
+	stopCh <-chan struct{},
```
Comment (wojtek-t) [inline staging/src/k8s.io/apiserver/pkg/admission/initializer/initializer.go:43]: @liggitt - I know there are a bunch of unit test failing now (due to a need to pass stopCh here in there). But before fixing them I would like to ensure that this approach is the desired one on how to implement it.

File: staging/src/k8s.io/apiserver/pkg/server/options/etcd.go [hunk 3]
```diff
@@ -226,7 +226,7 @@ func (s *EtcdOptions) ApplyWithStorageFactoryTo(factory serverstorage.StorageFac
 }
 
 func (s *EtcdOptions) addEtcdHealthEndpoint(c *server.Config) error {
-	healthCheck, err := storagefactory.CreateHealthCheck(s.StorageConfig, c.StopNotify())
```
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/server/options/etcd.go:229]: this doesn't look related to this PR, is this a separate fix?

Comment (wojtek-t) [reply staging/src/k8s.io/apiserver/pkg/server/options/etcd.go:229]: it is related - I wanted to have a single signal and for resource quota I want to stop it only after all requests are drained (which btw seems to be fine also in this case).

[But I'm happy to move it to a separate PR too.]

File: staging/src/k8s.io/apiserver/pkg/server/options/admission.go (hunk 1 above)
Comment (liggitt) [reply staging/src/k8s.io/apiserver/pkg/server/options/admission.go:147]: that's really subtle... how do we know this isn't breaking dependencies in the opposite direction?

File: staging/src/k8s.io/apiserver/pkg/server/config.go [hunk 4]
```diff
@@ -528,9 +528,9 @@ func completeOpenAPI(config *openapicommon.Config, version *version.Info) {
 	}
 }
 
-// StopNotify returns a lifecycle signal of genericapiserver shutting down.
-func (c *Config) StopNotify() <-chan struct{} {
-	return c.lifecycleSignals.ShutdownInitiated.Signaled()
+// DrainedNotify returns a lifecycle signal of genericapiserver already drained while shutting down.
+func (c *Config) DrainedNotify() <-chan struct{} {
+	return c.lifecycleSignals.InFlightRequestsDrained.Signaled()
```
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/server/config.go:533]: InFlightRequestsDrained is definitely clearer than ShutdownInitiated... is the swap correct for all existing callers / consumers?

File: staging/src/k8s.io/apiserver/pkg/server/options/admission.go (hunk 1 above)
Comment (wojtek-t) [reply staging/src/k8s.io/apiserver/pkg/server/options/admission.go:147]: I was looking into it and i didn't find anything suspicious.

The alternatively would be to adjust resource quota to also set stopCh of the underlying QuotaEvaluator, which fixes this particular problem, but doesn't address a potential future problems...

File: staging/src/k8s.io/apiserver/pkg/admission/plugin/resourcequota/admission.go [hunk 5]
```diff
@@ -82,21 +82,26 @@ type liveLookupEntry struct {
...
+func NewResourceQuota(config *resourcequotaapi.Configuration, numEvaluators int) (*QuotaAdmission, error) {
 	quotaAccessor, err := newQuotaAccessor()
 	if err != nil {
 		return nil, err
 	}
 
 	return &QuotaAdmission{
 		Handler:       admission.NewHandler(admission.Create, admission.Update),
-		stopCh:        stopCh,
+		stopCh:        nil,
```
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/admission/plugin/resourcequota/admission.go:93]: it's hard to reason about this being nil at init time and set to a non-nil thing later ... how do we prevent things from propagating handles to this further down into the plugin/controller/etc and not picking up the update in SetShutdownNotification?

File: staging/src/k8s.io/apiserver/pkg/server/config.go (hunk 4 above)
Comment (wojtek-t) [reply staging/src/k8s.io/apiserver/pkg/server/config.go:533]: There is exactly one user of that (this was added within last two weks) being etcd healthcheck, and it's correct there.
So yeah - we should swap - I will split that our into a separate PR, I guess.

File: staging/src/k8s.io/apiserver/pkg/admission/initializer/initializer.go (hunk 2 above)
Comment (liggitt) [reply staging/src/k8s.io/apiserver/pkg/admission/initializer/initializer.go:43]: hrmm, ideally this would be available at plugin init time, but that requires changing the signature of the plugin init function, which isn't great. asking the plugin to init things without a stop channel, then later request and replumb that down into everything it initialized is tricky

File: staging/src/k8s.io/apiserver/pkg/admission/initializer/initializer.go [hunk 6]
```diff
@@ -56,6 +59,11 @@ func (i pluginInitializer) Initialize(plugin admission.Interface) {
 		wants.InspectFeatureGates(i.featureGates)
 	}
 
+	// Second tell the plugin about shutdown signal, so it can pass it to further initializations.
```
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/admission/initializer/initializer.go:62]: shouldn't this be the first thing we call, so any initialization done anywhere else (like in InspectFeatureGates) has the stop channel to propagate?

File: staging/src/k8s.io/apiserver/pkg/admission/plugin/resourcequota/admission.go [hunk 7]
```diff
@@ -82,21 +82,26 @@ type liveLookupEntry struct {
...
 		numEvaluators: numEvaluators,
 		config:        config,
 		quotaAccessor: quotaAccessor,
 	}, nil
 }
 
+// SetShutdownNotification sets the stop channel into QuotaAdmission.
+func (a *QuotaAdmission) SetShutdownNotification(stopCh <-chan struct{}) {
+	a.stopCh = stopCh
```
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/admission/plugin/resourcequota/admission.go:102]: everywhere we're using stopCh, can we ensure it is non-nil?

File: staging/src/k8s.io/apiserver/pkg/server/config.go (hunk 4 above)
Comment (wojtek-t) [reply staging/src/k8s.io/apiserver/pkg/server/config.go:533]: Split this out to https://github.com/kubernetes/kubernetes/pull/110129

File: staging/src/k8s.io/apiserver/pkg/admission/initializer/interfaces.go [hunk 8]
```diff
@@ -49,6 +49,12 @@ type WantsQuotaConfiguration interface {
 	admission.InitializationValidator
 }
 
+// WantsShutdownNotification defines a function which sets the shutdown notification for admission plugins that need it.
+type WantsShutdownNotification interface {
```
[the remark this answers is not in the document]
Comment (wojtek-t) [reply staging/src/k8s.io/apiserver/pkg/admission/initializer/interfaces.go:53]: Done.

File: staging/src/k8s.io/apiserver/pkg/admission/plugin/resourcequota/admission.go (hunk 5 above)
Comment (wojtek-t) [reply staging/src/k8s.io/apiserver/pkg/admission/plugin/resourcequota/admission.go:93]: We don't have a generic mechanism for preventing propagating things down the stack before stopCh. But we don't even have a mechanism for setting stopCh now, so it's strict improvement.

I think it's mostly the same as for any regular bug - we need to catch that during the review.
In this PR I'm explicitly setting "initialization error" if we try to initialize something that requires stopCh and stopCh is unset, which will ensure that plugin will later not pass the validation.

File: staging/src/k8s.io/apiserver/pkg/admission/plugin/resourcequota/admission.go (hunk 7 above)
Comment (wojtek-t) [reply staging/src/k8s.io/apiserver/pkg/admission/plugin/resourcequota/admission.go:102]: Done (there is only one such place).

File: staging/src/k8s.io/apiserver/pkg/server/options/admission.go (hunk 1 above)
Comment (wojtek-t) [reply staging/src/k8s.io/apiserver/pkg/server/options/admission.go:147]: > (and ensure new reverse order dependencies don't get added)

The check that I added in resource quota that stopCh is actually set properly ensure that if this order will get changed, the initialization of resource quota plugin will start failing.
In general, every plugin that needs that, will have to handle it on their own.

**Review by wojtek-t — COMMENTED** [review]
@liggitt - thanks for the review - PTAL

File: plugin/pkg/admission/resourcequota/admission_test.go [hunk 9]
```diff
@@ -99,10 +99,11 @@ func createHandlerWithConfig(kubeClient kubernetes.Interface, informerFactory in
...
 
-	handler, err := resourcequota.NewResourceQuota(config, 5, stopCh)
+	handler, err := resourcequota.NewResourceQuota(config, 5)
 	if err != nil {
 		return nil, err
 	}
+	handler.SetDrainedNotification(stopCh)
 	handler.SetExternalKubeClientSet(kubeClient)
 	handler.SetExternalKubeInformerFactory(informerFactory)
 	handler.SetQuotaConfiguration(quotaConfiguration)
```
Comment (liggitt) [inline plugin/pkg/admission/resourcequota/admission_test.go:109]: can you switch these to using the initializer/validate pattern to make sure our tests match reality in terms in invocation order and that they are valid?

```go
	pluginInitializer := genericadmissioninitializer.New(...)
	pluginInitializer.Initialize(handler)
	err := admission.ValidateInitialization(handler)
  ...
```

File: test/integration/quota/quota_test.go [hunk 10]
```diff
@@ -68,18 +68,19 @@ func TestQuota(t *testing.T) {
...
 	admissionCh := make(chan struct{})
+	defer close(admissionCh)
 	clientset := clientset.NewForConfigOrDie(&restclient.Config{QPS: -1, Host: s.URL, ContentConfig: restclient.ContentConfig{GroupVersion: &schema.GroupVersion{Group: "", Version: "v1"}}})
 	config := &resourcequotaapi.Configuration{}
-	admission, err := resourcequota.NewResourceQuota(config, 5, admissionCh)
+	admission, err := resourcequota.NewResourceQuota(config, 5)
 	if err != nil {
 		t.Fatalf("unexpected error: %v", err)
 	}
+	admission.SetDrainedNotification(admissionCh)
```
Comment (liggitt) [inline test/integration/quota/quota_test.go:78]: same comment about using initializer and validating initialization

File: plugin/pkg/admission/resourcequota/admission_test.go (hunk 9 above)
Comment (wojtek-t) [reply plugin/pkg/admission/resourcequota/admission_test.go]: We unfortunately need more than one initializer (there are two in place).
That said - I switched to using that pattern here and in integration test.

**Review by wojtek-t — COMMENTED** [review]
@liggitt - thanks for the review, PTAL

File: plugin/pkg/admission/resourcequota/admission_test.go [hunk 11]
```diff
@@ -99,15 +101,19 @@ func createHandlerWithConfig(kubeClient kubernetes.Interface, informerFactory in
...
-	handler.SetQuotaConfiguration(quotaConfiguration)
 
-	return handler, nil
+	initializers := admission.PluginInitializers{}
+	initializers = append(initializers, genericadmissioninitializer.New(kubeClient, informerFactory, nil, nil, stopCh))
+	initializers = append(initializers, kubeapiserveradmission.NewPluginInitializer(nil, nil, quotaConfiguration))
+
+	for _, pluginInitializer := range initializers {
+		pluginInitializer.Initialize(handler)
+	}
```
Comment (liggitt) [inline plugin/pkg/admission/resourcequota/admission_test.go:115]: `initializers.Initialize(handler)`?

File: plugin/pkg/admission/resourcequota/admission_test.go [hunk 12]
```diff
@@ -99,15 +101,19 @@ func createHandlerWithConfig(kubeClient kubernetes.Interface, informerFactory in
...
 		return nil, err
 	}
-	handler.SetExternalKubeClientSet(kubeClient)
-	handler.SetExternalKubeInformerFactory(informerFactory)
-	handler.SetQuotaConfiguration(quotaConfiguration)
 
-	return handler, nil
+	initializers := admission.PluginInitializers{}
+	initializers = append(initializers, genericadmissioninitializer.New(kubeClient, informerFactory, nil, nil, stopCh))
+	initializers = append(initializers, kubeapiserveradmission.NewPluginInitializer(nil, nil, quotaConfiguration))
```
Comment (liggitt) [inline plugin/pkg/admission/resourcequota/admission_test.go:111]: non-blocking, but if you want, could simplify to

```go
initializers := admission.PluginInitializers{
  genericadmissioninitializer.New(...
  kubeapiserveradmission.NewPluginInitializer(...
}
```

Comment (wojtek-t) [reply plugin/pkg/admission/resourcequota/admission_test.go]: fixed

File: plugin/pkg/admission/resourcequota/admission_test.go (hunk 11 above)
Comment (wojtek-t) [reply plugin/pkg/admission/resourcequota/admission_test.go]: fixed

**Review by wojtek-t — COMMENTED** [review]
Thanks @liggitt for the review - PTAL

File: staging/src/k8s.io/apiserver/pkg/admission/plugin/resourcequota/admission.go [hunk 13]
```diff
@@ -67,6 +67,7 @@ type QuotaAdmission struct {
 	numEvaluators      int
 	quotaAccessor      *quotaAccessor
 	evaluator          Evaluator
+	initializationErr  error
 }
 
 var _ admission.ValidationInterface = &QuotaAdmission{}
```
Comment (aojea) [inline staging/src/k8s.io/apiserver/pkg/admission/plugin/resourcequota/admission.go:73]:
```suggestion
var _ admission.ValidationInterface = &QuotaAdmission{}
var _ = genericadmissioninitializer.WantsDrainedNotification(&QuotaAdmission{})
```

File: staging/src/k8s.io/apiserver/pkg/admission/plugin/resourcequota/admission.go [hunk 14]
```diff
@@ -110,11 +116,21 @@ func (a *QuotaAdmission) SetExternalKubeInformerFactory(f informers.SharedInform
...
+		a.initializationErr = fmt.Errorf("quota configuration configured between stop channel")
+		return
+	}
 	a.evaluator = NewQuotaEvaluator(a.quotaAccessor, a.quotaConfiguration.IgnoredResources(), generic.NewRegistry(a.quotaConfiguration.Evaluators()), nil, a.config, a.numEvaluators, a.stopCh)
 }
 
 // ValidateInitialization ensures an authorizer is set.
 func (a *QuotaAdmission) ValidateInitialization() error {
+	if a.initializationErr != nil {
+		return nil
```
Comment (aojea) [inline staging/src/k8s.io/apiserver/pkg/admission/plugin/resourcequota/admission.go:129]:
```suggestion
		return a.initializationErr 
```

Comment (aojea) [reply staging/src/k8s.io/apiserver/pkg/admission/plugin/resourcequota/admission.go:129]: should return an error, right?

File: staging/src/k8s.io/apiserver/pkg/admission/plugin/resourcequota/admission.go (hunk 13 above)
Comment (wojtek-t) [reply staging/src/k8s.io/apiserver/pkg/admission/plugin/resourcequota/admission.go:76]: done

File: staging/src/k8s.io/apiserver/pkg/admission/plugin/resourcequota/admission.go (hunk 14 above)
Comment (wojtek-t) [reply staging/src/k8s.io/apiserver/pkg/admission/plugin/resourcequota/admission.go]: dohh..

good catch - also added a test for that

**Comment by aojea** [comment]
lgtm , defer to Jordan
